WIP: update stable-diffusion.cpp to 5900ef6605c6 (new API)

wbruna opened the pull request:
This updates the embedded stable-diffusion.cpp code, and apply the necessary changes to the API adapter.

I've tried to preserve the local changes to the codebase:

* get_num_physical_cores got renamed to sd_get_num_physical_cores
* utf filenames on Windows
* a few header file changes for third party libs (json lib path, STB_IMAGE_ macros)
* logging / LOG_DEBUG changes
* automatic taesd model path
* fixed lora instead of enabling through the prompt
* kontext and photomaker images passed directly
* auto-disabling photomaker for non-SDXL models
* the fix for VAE tiling already applied
* something else I'm probably forgetting

I've applied other minor changes, like formatting, to minimize the diff against sd.cpp mainline.

Right now, the code merely builds successfully, but I'll keep working on it over the next few days.

wbruna committed:
Update stable-diffusion.cpp to 5900ef6605c6 (new API)

LostRuins commented:
thanks. i think most of the files should be relatively easy to upgrade, except for `stable-diffusion.cpp` itself which has a few key points of divergence mostly to do with disk access and image data passing.

LostRuins commented on otherarch/sdcpp/stable-diffusion.cpp:
```diff
@@ -193,80 +174,84 @@ class StableDiffusionGGML {
+        if (strlen(SAFE_STR(sd_ctx_params->vae_path)) > 0) {
+            LOG_INFO("loading vae from '%s'", sd_ctx_params->vae_path);
+            if (!model_loader.init_from_file(sd_ctx_params->vae_path, "vae.")) {
+                LOG_WARN("loading vae from '%s' failed", sd_ctx_params->vae_path);
```
have to be very careful in this file, e.g. my `endswithsafetensors` code looks to be affected in this draft?

there are a bunch of changes - I'd recommending `diff` ing the current koboldcpp version against files in https://github.com/leejet/stable-diffusion.cpp/tree/b1cc40c35cec90420d15e9e25bd34f4f85ac0c5d

wbruna replied:
> have to be very careful in this file, e.g. my endswithsafetensors code is already broken in the current draft. likewise with LoRA.

So it falls back to treat the file as a separate diffusion model. Yeah, I did miss that chunk. 

Speaking of LoRA, is the `pending_apply_lora_fname` code a leftover? I see no calls to `set_pending_lora` apart from commented out code.

LostRuins replied:
Yeah, that has been replaced by a direct LoRA load in sdcpp adapter 
<img width="819" height="226" alt="image" src="https://github.com/user-attachments/assets/9fee9251-c947-4a9e-9428-de2ffdbd8afe" />

wbruna replied:
Shall I keep the `set_pending_lora` code around, to minimize the changes in this update? Or clean it up, to minimize the changes against upstream? (either approach is fine by me)

LostRuins replied:
For this case it doesn't matter, you can remove it if you like. Just make sure all current functional code is kept

LostRuins commented on otherarch/sdcpp/lora.hpp:
```diff
@@ -3,7 +3,7 @@
 #include "ggml_extend.hpp"
 
-#define LORA_GRAPH_SIZE 20480
+#define LORA_GRAPH_BASE_SIZE 10240
```
this value is known to crash

wbruna replied:
Current upstream applies an heuristic on top of that number: https://github.com/leejet/stable-diffusion.cpp/commit/1ce1c1adcadb30d112e812cd5ce4cf210f968bc0 (that new code is included on line 241 below).

LostRuins replied:
hmm is that enough tho? I have not tested it extensively, I only know the old one does crash. I'd personally bump that number unless you are very confident it's high enough in all cases. (its negligible size anyway)

wbruna replied:
Well, I can't say I fully understand that heuristic, but I did some tests back when leejet applied that change, and the final size actually got even larger for known problematic LoRAs (around 33000 for the dmd2).

wbruna committed:
Clean up pending LoRA code and simplify LoRA changes to upstream

wbruna committed:
Move VAE tiling disabling for TAESD to sdtype_adapter.cpp

wbruna committed:
Move auxiliary ctx functions to sdtype_adapter.cpp

wbruna commented:
The current code seems to be working for basic gens + LoRAs, except for an upstream issue with img2img + flash attention ( [leejet/stable-diffusion#756](https://github.com/leejet/stable-diffusion.cpp/issues/756) ). I still need to test the more extensively changed code paths (Kontext, Photomaker, TAESD, ...).

I've tried to move some local changes from stable-diffusion.cpp to sdtype_adapter.cpp, to facilitate future updates. The Chroma detection got a bit more complex, so let me know if you prefer the previous approach.

For removing the LoRAs from the prompt, I've opted for emptying the LoRA list instead, to make the diff cleaner. That also makes a nice place to print a warning, in case the user tries to include a LoRA through the prompt.

It may be worth to change the Kontext and Photomaker image lists to use a single image reference list, instead of one list per model. That could avoid further interface changes to support other edit models in the future, like Instruct-Pix2pix and CosXL. But of course that could be done after this update.

wbruna commented:
Chroma isn't working with flash attention (black images), but could be that same upstream issue. Maybe the safest would be a separate flash attention control for image generation... I can add one later, along with configs for the new conv2d_direct optimizations.

I've also included here the fix for #1672 , to facilitate the tests. Apart from that: SD1.5, SDXL, SD3.5, Chroma, Flux, Kontext, Photomaker, inpaint, LoRAs and TAESD seem to be working so far (Linux + Vulkan).

wbruna committed:
Use ref_images parameter for Kontext images

LostRuins commented:
I decided to test flash attention on chroma in v1.97.1 (latest release) cuda, and I also got a black square, so at least that (appears) to be the case for now.

wbruna commented:
I've rebased into 1.97.1, and folded missing stuff into the original commit, still keeping the refactors separate in case we need to tweak them.

I could add a workaround to disable flash attention for Chroma. For img2img / inpaint is trickier, since the flag is set at model loading time; but maybe it could be changed inside the ctx at inference time, like the VAE tiling flag.

LostRuins commented:
we can do enhancements as a separate PR, lets get full feature parity and no regressions first.

wbruna commented:
I took a look at enabling flash attention at inference time. It's _doable_, but we'd need to touch a lot of objects on the model tree to propagate the flag. Unless upstream is willing to switch to that approach (as was done for the enable_conv2d_direct flag), it may be best to simply disable flash attention for image gen until https://github.com/leejet/stable-diffusion.cpp/issues/756 moves forward.

So, I believe the current code is ready to be reviewed.

LostRuins commented on otherarch/sdcpp/clip.hpp:
```diff
@@ -545,9 +545,9 @@ class CLIPEmbeddings : public GGMLBlock {
-        enum ggml_type position_wtype = GGML_TYPE_F32;  //(tensor_types.find(prefix + "position_embedding.weight") != tensor_types.end()) ? tensor_types[prefix + "position_embedding.weight"] : GGML_TYPE_F32;
+    void init_params(struct ggml_context* ctx, const String2GGMLType& tensor_types = {}, const std::string prefix = "") {
+        enum ggml_type token_wtype    = GGML_TYPE_F32;
+        enum ggml_type position_wtype = GGML_TYPE_F32;
```
Edit: i'm not sure if we should revert these lines back to the previous state in the patch i applied https://github.com/LostRuins/koboldcpp/commit/cbdee99354753ac93ee2a37299ac4cf74c02f614 

Reason:  full GGML_TYPE_F32 consumes a good bit more VRAM prevents running image+text model on colab. 

But if upstream expects them to be f32 in future, keeping the modification would also be not ideal. 

So maybe we can stick to F32. I'd just have to offload fewer layers.

wbruna replied:
All right, Ill leave like this for now.
(making a note for myself: that was from a fix by stdhupf , specifically for quantized models and LoRA: https://github.com/leejet/stable-diffusion.cpp/commit/fbd42b6fc16d14fbd362993fa1d083740a05f113)

wbruna replied:
Reading the comments on that issue, the problem seems specific to embeddings. But I'm not sure if it's only for textual inversion files (which Koboldcpp doesn't support anyway), or any kind of embedding (which would include eg. Photomaker stuff).

I could restore the workaround, then test it again to see if I can reproduce that issue.

wbruna replied:
I investigated this a bit:
* restoring the old behavior does reduce peak memory usage, around 190MB for a SDXL inference (6.60Gb -> 6.41Gb);
* I get garbage images with the old code, but only if I try to use a textual inversion file. A simple generation is seemingly unaffected;
* that code was introduced on https://github.com/leejet/stable-diffusion.cpp/commit/7ce63e740cc01715d075b5520880712da3a9f9aa , and originally used the tensor type unconditionally. So, looks like textual inversions already didn't work back then, at least on SDXL.

I'll restore the workaround, and see if upstream would accept a patch to enable it when the inference isn't using textual inversions.

LostRuins replied:
Alright, i'll leave the choice up to you - we can use the F32 if you think it's better. Users will just have to allocate more memory.

Alternatively, we can make the F32 conditional on having embeddings/textual inversion (anything that causes crashes). I am fine with that.

wbruna replied:
I redid a bunch of tests disabling the F32 conversion, and couldn't find a situation where it caused a failed generation on Koboldcpp (Photomaker embeddings work fine). And it'd also be one less behavior change for this update, so I think it's worth including for now.

I'll prepare a PR for upstream; we can drop the workaround if/when it gets merged.

LostRuins commented on otherarch/sdcpp/model.cpp:
```diff
@@ -1964,20 +2091,21 @@ bool ModelLoader::load_tensors(on_new_tensor_cb_t on_new_tensor_cb, ggml_backend
-            ++tensor_count;
-            if(tensor_count<2 || tensor_count%5==0 || (tensor_count+10) > processed_tensor_storages.size())
-            {
-                //throttle progress printing
```
you've accidentally removed my progress printing throttling code

wbruna replied:
Oops, that one got tangled with another change.

Edit: fixed.

LostRuins commented on otherarch/sdcpp/sdtype_adapter.cpp:
```diff
@@ -160,6 +160,10 @@ bool sdtype_load_model(const sd_load_model_inputs inputs) {
     {
         taesdpath = executable_path + "taesd.embd";
         printf("With TAE SD VAE: %s\n",taesdpath.c_str());
+        if (cfg_tiled_vae_threshold < 8192) {
```
hmm where does this come from? what is it for? looks like it was added by mistake

wbruna replied:
Ah, that's from me; I should've added a comment. `cfg_tiled_vae_threshold < 8192` is a way to test if VAE tiling is active for some resolution (the point is disabling tiling for TAESD outside stable-diffusion.cpp ).

LostRuins replied:
Hmm alright. So this actually disables VAE tiling when TAESD is on if I understand it correctly?

LostRuins commented on otherarch/sdcpp/sdtype_adapter.cpp:
```diff
@@ -437,6 +444,27 @@ static void sd_fix_resolution(int &width, int &height, int img_hard_limit, int i
+    return ctx->sd->version;
+}
+
+bool loaded_model_is_chroma(sd_ctx_t* ctx)
```
I trust you have tested this can differentiate between all flux (schnell/dev) vs chroma? Previously I checked for presence of `distilled_guidance_layer.in_proj.weight` but I am fine with any approach that works.

wbruna replied:
I'm actually 'borrowing' a flag from upstream that is set with that same test (at the `FluxRunner` constructor), and it's used to differentiate Chroma from other Flux variants throughout the code.

I've tested Chroma, Schnell and Kontext. But I'm remembering there's a check that only avoids bad inference parameters, and I could have set good ones myself; I'll double check.

LostRuins replied:
Thanks!

LostRuins commented on otherarch/sdcpp/stable-diffusion.cpp:
```diff
@@ -1267,8 +1356,7 @@ class StableDiffusionGGML {
         } else {
-            //koboldcpp never use tiling with taesd
-            if (false && vae_tiling) {  // TODO: support tiling vae encode
+            if (vae_tiling) {
```
is this intentional? Iirc tiling breaks taesd

wbruna replied:
Yes, I've moved that to sdtype_adapter.cpp (around that check for `cfg_tiled_vae_threshold < 8192`). Anyway, that breakage was likely fixed by stdhupf's tiled VAE changes, so it may be worth testing it again. Do you remember if that happened in general, or on a specific model version?

LostRuins replied:
Can't remember. I will have to test again, it was a long time ago.

wbruna replied:
I think we may leave it like this for now, and test it again after the update.

LostRuins commented on otherarch/sdcpp/stable-diffusion.cpp:
```diff
@@ -1458,14 +1747,14 @@ sd_image_t* generate_image(sd_ctx_t* sd_ctx,
     }
+
+    int64_t t0 = ggml_time_ms();
+    sd_ctx->sd->apply_loras(lora_f2m);
```
`sd_ctx->sd->apply_loras(lora_f2m);`
why do we need apply_loras, kcpp is not allowed to load loras at runtime

wbruna replied:
That just minimizes the changes against upstream. All LoRAs were removed from that list on the code right above that.

LostRuins commented on otherarch/sdcpp/stable-diffusion.cpp:
```diff
@@ -193,80 +170,101 @@ class StableDiffusionGGML {
 
         LOG_INFO("Version: %s ", model_version_to_str[version]);
 
-        if(id_embeddings_path!="" && version!=VERSION_SDXL)
```
is removing this intentional?
edit: i see you moved it lower.

wbruna committed:
Workaround for flash attention with img2img

leejet/stable-diffusion.cpp#756

wbruna committed:
Workaround for Chroma with flash attention, debug prints

wbruna committed:
Drop clip skip workaround (fixed upstream)

wbruna commented:
I've fixed the pretty_progress throttling, added workarounds for the flash attention issues with Chroma and img2img, and cleaned up the workaround for clip_skip that got fixed upstream. I don't have much time right now to do full tests (Flux is really slow here :-) ), but I'll do it later.

LostRuins commented:
Thanks for everything so far.

wbruna committed:
Disable forcing CLIP weights to F32 for reduced memory usage

wbruna commented:
I redid the tests with the CLIP weights change, and just found a minor bug on the pretty_progress throttling. ROCm is working fine, too. As far as I can tell, it's ready!

LostRuins commented:
Tested Sd1.5, SDXL, Flux, Kontext, seem ok so far.

Are you able to use SD3.5 files? I think some time back there was a regression that messed up sd3.5 for me, I also get a black square the latest release either, so it's not the fault of this PR.

Testing on this file https://huggingface.co/Comfy-Org/stable-diffusion-3.5-fp8/blob/main/sd3.5_medium_incl_clips_t5xxlfp8scaled.safetensors


It does work back in 1.78 when it was introduced. 
1.80.3 is still working. 
It is broken by 1.81.1
Trying to figure out when and why it broke.

Edit: looks like it broke at 2a890ec25ae403259c6816878457c52c8a4a96f0

wbruna commented:
At least SD3.5 medium is working for me; [sd3.5m_turbo-Q4_K_M.gguf](https://huggingface.co/tensorart/stable-diffusion-3.5-medium-turbo/blob/main/sd3.5m_turbo-Q4_K_M.gguf) :

<img width="384" height="384" alt="test_sd3_turbo" src="https://github.com/user-attachments/assets/c9c570ad-9250-4e65-8800-29d9bc758b53" />

and [sd3.5_medium-Q8_0.gguf](https://huggingface.co/city96/stable-diffusion-3.5-medium-gguf/blob/main/sd3.5_medium-Q8_0.gguf):

<img width="384" height="384" alt="test_sd3_medium" src="https://github.com/user-attachments/assets/6241922b-e3e9-4976-83ed-b566f8df4220" />

My sd3.5-large gguf is failing right now, but it's a failure to load, not garbage rendering; probably a messed up config.

Edit: using [this VAE](https://huggingface.co/calcuis/sd3.5-large-gguf/blob/7f72f2a432131bba82ecd1aafb931ac99f0f05f7/diffusion_pytorch_model.safetensors) (I don't remember why I picked this one, I just found it again by the SHA256 hash).

wbruna commented:
I can't get [SD3.5-Large-Turbo-GGUF-mixed-sdcpp](https://huggingface.co/stduhpf/SD3.5-Large-Turbo-GGUF-mixed-sdcpp) to load on Koboldcpp. And it's working fine on sd.cpp:

<img width="384" height="384" alt="test_sd3_large_sdcpp" src="https://github.com/user-attachments/assets/9d6be3e2-85ce-4536-a1d8-b1fcbc2b755b" />

I think Koboldcpp was falling back to another previously working configuration during my tests :-(

(also downloading the `sd3.5_medium_incl_clips_t5xxlfp8scaled.safetensors` file to test it here)

LostRuins commented:
When I use taesd i get a brown square, otherwise I get a black square 
https://github.com/leejet/stable-diffusion.cpp/issues/560

Edit: this was on medium

LostRuins reviewed (approved):
approving first, lets see if we can figure it out. 
if not, we can still merge it since all other stuff seems good to go.

LostRuins commented:
okay looking at your gguf example I think the fallback code is failing, so the tensor names are appended twice 
`model.diffusion_model.model.diffusion_model`...
let me test in an older koboldcpp first

I think the solution is to check for such tensor names before allowing the prefix to be appended.

wbruna commented:
I did a few tests with sd3.5_medium_incl_clips_t5xxlfp8scaled on sd.cpp:
* rocm, plain model: black image
* rocm, model + external vae: black image
* rocm, model + external vae + external t5xxl: works
* vulkan, model + external t5xxl: works

The external t5xxl comes from [t5xxl-Q4_K.gguf](https://huggingface.co/second-state/FLUX.1-dev-GGUF/blob/main/t5xxl-Q4_K.gguf).

Edit: [t5xxl-Q8_0.gguf](https://huggingface.co/second-state/FLUX.1-dev-GGUF/blob/d95134d322d31f272313e3ca9a6cbb20c77b7487/t5xxl-Q8_0.gguf) works fine too.

LostRuins commented:
@wbruna alright after I added my fix for the tensor names, I was able to load the model.

However, the results are very weird. I don't know if I am doing it wrong or if I used bad files, since **I loaded my own T5 and clip**, using your [SD3.5-Large-Turbo-GGUF-mixed-sdcpp](https://huggingface.co/stduhpf/SD3.5-Large-Turbo-GGUF-mixed-sdcpp). 

Can you merge my changes from https://github.com/LostRuins/koboldcpp/commit/7b5cf7143fd29608e87445d366a88546a5a81a92 and check if it works fine for you? 
Prompt: Cat with Hat

<img width="636" height="594" alt="image" src="https://github.com/user-attachments/assets/1bc492cf-8f40-4e00-ace2-d3d82428f311" />

I think maybe i have a faulty t5, clip or vae. but if it works for you i think thats good enough. please merge and try my fix

LostRuins commented:
> vulkan, model + external t5xxl: works

could you also check in sdcpp
vulkan, model + internal t5xxl (all in one, this fails for me.)

wbruna commented:
> I think maybe i have a faulty t5, clip or vae. but if it works for you i think thats good enough. please merge and try my fix

It works! This was with the just-pushed merge 9c039b2 , on Vulkan:

<img width="512" height="512" alt="test_sd35m_kobold" src="https://github.com/user-attachments/assets/f2e2cc11-2340-4ee1-8170-4485a94f4cf3" />

LostRuins commented:
Alright, if no other issues I will merge this now

wbruna commented:
> could you also check in sdcpp
> vulkan, model + internal t5xxl (all in one, this fails for me.)

Black image for me too.

wbruna commented:
@LostRuins , the target `sdmain` is currently failing to build. I was able to fix a few issues, but then I got `undefined reference to stbi_load`; and when I tried to fix _that_, I noticed conflicts between `otherarch/sdcpp/thirdparty` and `vendor/stb` due to version changes. Perhaps it'd be best to use only a single copy in `vendor/stb`?

LostRuins commented:
Alright I fixed the issues I could fine and it compiles fine for me. Please pull latest experimental. Does it work for you now?

wbruna commented:
Yeah, it builds fine now; thanks!

LostRuins commented:
Btw i've been keeping my eye on the upstream Qwen Image and WAN developments. Thanks to your PR kcpp is now in sync with the latest API, so merging it shouldn't be too hard. 

However, we'd probably want to think of a good approach for the API and frontend side since video files will no doubt be massive. AVI is probably not the best format due to it's size, though transcoding without FFMPEG (which cannot be used) will be tricky. Open to suggestions if you have any.

wbruna commented:
> Btw i've been keeping my eye on the upstream Qwen Image and WAN developments. Thanks to your PR kcpp is now in sync with the latest API, so merging it shouldn't be too hard.

Yeah... The main issue is the needed ggml support; I don't know how close leejet is to upstreaming them. And I'm already missing fixes on that branch (he's working on a version with that slow Vulkan build time 😢). Unrelated breakage could be an issue, too... But I suspect leejet already had wan in mind for that big refactor, so it could be less of an issue now.

And I'm considering again that old idea of keeping an sd branch as an 'upstream' for Kobold. Two, actually: one branch just to pick still-not-applied upstream PRs, another with the specific tweaks for Koboldcpp. I kinda did that locally for the VAE tiling fixes, but making it 'official' would make the changes easier to track (since it doesn't look like we'll have those applied to sd master any time soon...).

> However, we'd probably want to think of a good approach for the API and frontend side since video files will no doubt be massive. AVI is probably not the best format due to it's size, though transcoding without FFMPEG (which cannot be used) will be tricky. Open to suggestions if you have any.

Koboldcpp can't embed libavcodec? (honest question; I know very little about video encoding. I confess I'm afraid of even trying video gen on my poor card 😅)

LostRuins commented:
It's really big cause it supports dozens of codecs. And it has multiple mixed licenses, LGPL is technically usable within APGL but i'd prefer to avoid it if i can. I only need ONE single codec that can be played in a browser, thats it.

ehoogeveen-medweb commented:
Would it be possible to use the CLI if it's available, or would that be too indirect? I already have ffmpeg on my PATH anyway for yt-dlp, for example.

LostRuins commented:
Most likely if I cannot find something suitable the output will remain as MJPEG, which the clients can then convert on their side to whatever format they prefer

wbruna commented:
> It's really big cause it supports dozens of codecs. And it has multiple mixed licenses, LGPL is technically usable within APGL but i'd prefer to avoid it if i can. I only need ONE single codec that can be played in a browser, thats it.

Maybe OpenH264? It's BSD licensed, and the binaries seem relatively small:
https://github.com/cisco/openh264/releases/tag/v2.6.0

wbruna commented:
Wan support just got merged into master. There is a ton of changes, so it may be reasonable to just update the sd.cpp code for now, and add video gen support later.

Code-wise, the main issue is that the VAE tiling fixes conflict with some of the Wan changes; I asked stduhpf to update the PR. main.cpp also got a lot of changes, but the API itself didn't change that much, so sdtype_adapter may not need too many fixes right now.

LostRuins commented:
Alright cool.

wbruna commented:
I think I've found a way to better track those changes. I'm keeping a branch with the upstream code + pending PRs, which will act as a vendor branch:

https://github.com/wbruna/stable-diffusion.cpp/tree/koboldcpp_sd_base

And another branch with the needed local changes:

https://github.com/wbruna/stable-diffusion.cpp/tree/koboldcpp_sd_changes

The idea is to keep `koboldcpp_sd_base` up-to-date with upstream first, then merge it into `koboldcpp_sd_changes`.

Right now, that merge is just a first try; I'll review it later, and open a PR. Possibly including those FA changes that just arrived on master...
